package/MDAnalysis/lib/pkdtree.py

```diff
@@ -0,0 +1,181 @@
+    case. For instance, if the wrapped center point lies well within the cell
+    and far away from the cell boundaries, there may be no need to generate
+    any center point image.
+    """
+
+    def __init__(self, box, bucket_size=10):
+        """
+        Parameters
+        ----------
+        box : array-like or ``None``, optional, default ``None``
+          Simulation cell dimensions in the form of
+          :attr:`MDAnalysis.trajectory.base.Timestep.dimensions` when
+          periodic boundary conditions should be taken into account for
+          the calculation of contacts.
+        bucket_size : int
+          Number of entries in leafs of the KDTree. If you suffer poor
+          performance you can play around with this number. Increasing the
+          `bucket_size` will speed up the construction of the KDTree but
+          slow down the search.
+        """
+        self.dim = 3  # Strict implementation for 3D
+        self.kdt = _CKDTree.KDTree(self.dim, bucket_size)
+        self.built = 0
+        if len(box) != self.dim:
+            raise Exception('Expected array of length {}'.format(self.dim))
+        self.box = np.copy(np.asarray(box))
+        # zero is the flag for no periodic boundary conditions
+        self.box = np.where(self.box == np.inf, 0.0, self.box)
+        self.box = np.where(self.box > 0.0, self.box, 0.0)
+        if not self.box.any():
+            raise Exception('No periodic axis found')
+        self._indices = list()
+
+    def set_coords(self, coords):
+        """
+        Add coordinates of the points. Wrapping of coordinates to the central
+        cell is enforced along the periodic axes.
+
+        Parameters
+        ----------
+        coords: NumPy.array
+          Positions of points, shape=(N, 3) for N atoms.
+        """
+        if coords.min() <= -1e6 or coords.max() >= 1e6:
+            # Same exception class as in Bio.KDTree.KDTree.set_coords
+            raise Exception('Points should lie between -1e6 and 1e6')
+        if len(coords.shape) != 2 or coords.shape[1] != self.dim:
+            raise Exception('Expected a (N, {}) NumPy array'.format(self.dim))
+        wrapped_data = (coords - np.where(self.box > 0.0,
+                                          np.floor(coords / self.box) * self.box, 0.0))
+        self.kdt.set_data(wrapped_data)
+        self.built = 1
+
+    def find_centers(self, center_point, radius):
+        """
+        Find relevant images of a center point.
+
+        Parameters
+        ----------
+        center_point: NumPy.array
+          Coordinates of the query center point
+        radius: float
+          Maximum distance from center in search for neighbors
+
+        Returns
+        ------
+        :class:`List`
+          center point and its relevant images
+        """
+        wrapped_c = (center_point - np.where(self.box > 0.0,
+                     np.floor(center_point / self.box) * self.box, 0.0))
```

Comment:
sure this check is needed every time. I ensured the validity of `self.box` once in the beginning already.

Reply:
@kain88-de I don't understand this comment. `find_centers` is called only one time for a given query `center_point`. Can you expand on this?

Reply:
You are using the pattern `np.where(self.box > 0.0, if_true, if_false)`  several times outside of the `__init__` function. In the `__init__` you are already checking that `self.box` is formatted to your liking. Therefore these tests should be not needed any more.

Reply:
`self.box[i]` is zero if the system is non-periodic along the `i`th axis, so I have to use `np.where(self.box > 0.0...` to prevent division by zero in `center_point / self.box` for index `i`. Please let me know if there's a simpler rewrite for this line.

Reply:
I'll have a look. One way to write this in python though is normal to use the [ternary operator](http://www.cplusplus.com/articles/1AUq5Di1/)

```python
true_case if cond else false_case
```

I actually wasn't aware that numpy's `where` function can be used as a ternary operator.